.github/workflows/make-and-test.yml

```diff
@@ -106,7 +194,34 @@ jobs:
           pip install ipykernel
           python -m ipykernel install --prefix=./jupyter-local --name python3-local --display-name "Python 3 (Local)"
 
-      - run: cd src && npm install -g pnpm
-      - run: cd src && pnpm run make
-      - run: source venv/bin/activate && cd src && pnpm run test || pnpm run test || pnpm run test
-      - run: cd src && pnpm run depcheck
+      - name: install pnpm
+        #run: cd src && npm install -g pnpm
+        uses: pnpm/action-setup@v4
+        with:
+          version: 10
+          run_install: false
+
+      - name: Restore build artifacts
+        uses: actions/cache@v4
+        with:
+          path: "src" # src/**/dist
+          key: ${{ runner.os }}-build-${{ matrix.pg-version }}-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-build-${{ matrix.pg-version }}-
+
+      # - name: Download build artifacts
+      #   uses: actions/download-artifact@v4
+      #  with:
+      #    name: build-output
+
+      - name: Run tests
+        run: |
+          source venv/bin/activate
+          cd src/packages/${{ matrix.dir }}
+          if [ ! -f package.json ]; then
+            echo "::warning::No package.json found in ${{ matrix.dir }}"
+          elif ! jq -e '.scripts.test' package.json > /dev/null; then
+            echo "::notice::No test script in ${{ matrix.dir }}"
+          fi
+          pnpm run test
```

Comment:
Just use `pnpm run --if-present test` instead of jq shell commands.  

https://pnpm.io/cli/run#--if-present

Also, do we have packages with no package.json?